src/main/scala/org/broadinstitute/dsde/workbench/sam/api/SecurityDirectives.scala

```diff
@@ -18,30 +19,32 @@ trait SecurityDirectives {
-        if (hasAccess(requestedActions, actions)) innerRoute
-        else if (actions.isEmpty)
-          Directives.failWith(
-            new WorkbenchExceptionWithErrorReport(
-              ErrorReport(StatusCodes.NotFound, s"Resource ${resource.resourceTypeName.value}/${resource.resourceId.value} not found")))
-        else
-          Directives.failWith(
-            new WorkbenchExceptionWithErrorReport(ErrorReport(
-              StatusCodes.Forbidden,
-              s"You may not perform any of ${requestedActions.mkString("[", ", ", "]").toString.toUpperCase} on ${resource.resourceTypeName.value}/${resource.resourceId.value}"
-            )))
+      onSuccess(hasPermissionOneOf(resource, requestedActions, userId)) { hasPermission =>
+        if (hasPermission) {
+          innerRoute
+        } else {
+
+          // in the case where we don't have the required action, we need to figure out if we should return
+          // a Not Found (you have no access) vs a Forbidden (you have access, just not the right kind)
+          onSuccess(policyEvaluatorService.listResourceAccessPoliciesForUser(resource, userId)) { policies =>
+
+            if (policies.isEmpty) {
+              Directives.failWith(
+                new WorkbenchExceptionWithErrorReport(
+                  ErrorReport(StatusCodes.NotFound, s"Resource ${resource.resourceTypeName.value}/${resource.resourceId.value} not found")))
+            } else {
+              Directives.failWith(
+                new WorkbenchExceptionWithErrorReport(ErrorReport(
+                  StatusCodes.Forbidden,
+                  s"You may not perform any of ${requestedActions.mkString("[", ", ", "]").toString.toUpperCase} on ${resource.resourceTypeName.value}/${resource.resourceId.value}"
+                )))
+            }
+          }
+        }
       }
     }
 
-  private def listActions(resource: FullyQualifiedResourceId, userId: WorkbenchUserId, requestedActions: Set[ResourceAction]): IO[Set[ResourceAction]] =
-    // this is optimized for the case where the user has permission since that is the usual case
-    // if the first attempt shows the user does not have permission, force a second attempt
-    for {
-      actionsAttempt1 <- policyEvaluatorService.listUserResourceActions(resource, userId, force = false)
-      actionsAttempt2 <- if (hasAccess(requestedActions, actionsAttempt1)) IO.pure(actionsAttempt1)
-      else policyEvaluatorService.listUserResourceActions(resource, userId, force = true)
-    } yield actionsAttempt2
-
-  private def hasAccess(requestedActions: Set[ResourceAction], actions: Set[ResourceAction]): Boolean =
-    actions.intersect(requestedActions).nonEmpty
+  private def hasPermissionOneOf(resource: FullyQualifiedResourceId, actions: Set[ResourceAction], userId: WorkbenchUserId): IO[Boolean] =
+    actions.toList.existsM(policyEvaluatorService.hasPermission(resource, _, userId))
+  
```

Comment:
for consideration ... would it be more optimal to try 
```
actions.toList.existsM(policyEvaluatorService.hasPermissionShallowCheck(resource, _, userId))
```
first, and iif that is false, then try
```
actions.toList.existsM(policyEvaluatorService.hasPermissionFullCheck(resource, _, userId))
```
?

Additionally, consider changing these directives to accept an `Iterable[ResourceAction]` instead of `Set[ResourceAction]`, which would allow callers of these directives to optimize by listing the most-likely action first in the iteration order.

Reply:
great points, will incorporate

Reply:
ah, and if you do  the `listUserResourceActions` as part of the permission check, you can reuse its result for figuring out 403 vs 404